src/main/java/com/paloaltonetworks/osc/api/PANManagerSecurityGroupApi.java

```diff
@@ -1,110 +1,228 @@
+
+    public static final String SG_TAG_PREFIX = "OSC_SecurityGroup_";
+
+    private VirtualSystemElement vs;
+    private ApplianceManagerConnectorElement mc;
+    private PanoramaApiClient panClient;
+
+    private String addrXpath;
+    private String tagXpath;
+    private String devGroup;
+
+    public PANManagerSecurityGroupApi(ApplianceManagerConnectorElement mc, VirtualSystemElement vs,
+            PanoramaApiClient panClient) {
+        this.vs = vs;
+        this.mc = mc;
+        this.panClient = panClient;
+
+        this.devGroup = devGroupName(vs.getId());
+        this.addrXpath = String.format(PanoramaApiClient.XPATH_DEVGROUP_TEMPL + "/address", this.devGroup);
+        this.tagXpath = String.format(PanoramaApiClient.XPATH_DEVGROUP_TEMPL + "/tag", this.devGroup);
+    }
+
+    /**
+    *
+    * Note that the order of arguments is reversed: {@code updateSecurityGroup} has {@code name} as a first argument
+    * and {@code id} as a second.
+    *
+    * @see ManagerSecurityGroupApi
+    *
+    */
+    @Override
+    public String createSecurityGroup(String name, String sgId, SecurityGroupMemberListElement memberList)
+            throws Exception {
+        try {
+            LOG.info("Adding security group {} with members {}", name, memberList);
+            doUpdateSecurityGroup(sgId, name, memberList);
+        } catch (Exception e) {
+            LOG.error("Exception adding security group " + name, e);
+            return null;
+        }
+
+        return sgId;
+    }
+
+    /**
+     *
+     * Note that the order of arguments is reversed: {@code createSecurityGroup} has {@code name} as a first argument
+     * and {@code id} as a second.
+     *
+     * @see ManagerSecurityGroupApi
+     *
+     */
+    @Override
+    public void updateSecurityGroup(String sgId, String name, SecurityGroupMemberListElement memberList)
+            throws Exception {
+        try {
+            doUpdateSecurityGroup(sgId, name, memberList);
+        } catch (Exception e) {
+            LOG.error("Exception updating security group " + sgId, e);
+        }
+    }
+
+    @Override
+    public void deleteSecurityGroup(String sgId) throws Exception {
+        String sgTag = securityGroupTag(sgId);
+        Set<String> ipsOnMgr = fetchMgrIpsByTag(sgTag);
+        try {
+            for (String ip : ipsOnMgr) {
+                LOG.info("Deleting address {}", ip);
+                deleteAddress(ip);
+            }
+        } catch (Exception e) {
+            LOG.error("Exception deleting security group " + sgId, e);
+            throw e;
+        }
+
+        deleteSGTag(sgId);
+        this.panClient.configCommit();
+    }
+
+    @Override
+    public List<? extends ManagerSecurityGroupElement> getSecurityGroupList() throws Exception {
+        return null;
+    }
+
+    @Override
+    public ManagerSecurityGroupElement getSecurityGroupById(String sgId) throws Exception {

```

Comment:
iterate on the getSecurityGroupList to return the element. Or if there is specific api to get the security group tag, use that.